### DISCUSSION TITLE
k8s crd to pkl generation

### DISCUSSION THREAD

**Discussion by Pythoner6** [apple/pkl#100] [Q&A] [answered]
Is there any existing way to generate pkl from a k8s crd? Maybe I've missed something but it looks like the current k8s packages can really only generate pkl from the k8s openapi spec, and not crds. I'm very interested in trying out pkl for configuring my k8s cluster but without support for crds that's a hard sell.

**Comment by holzensp** [comment, marked as answer]
You can generate the CRDs thanks to @jackkleeman's awesome work.

Docs here: https://pkl-lang.org/package-docs/pkg.pkl-lang.org/pkl-pantry/k8s.contrib.crd/current/generate/index.html

The simplest usage is:

```bash
pkl eval package://pkg.pkl-lang.org/pkl-pantry/k8s.contrib.crd@1.0.0#/generate.pkl \
  -m . \
  -p source="https://path/to/crd.yml"
```

This will generate Pkl files in the current directory for the provided CRD specs.

You can fine-tune this by creating your _own_ generate script that amends the published generator. For example, this causes the generator to use `module ResourceRequirements` for the `spec.resources` property of your CRD:

_generate.pkl_
```ts
amends "package://pkg.pkl-lang.org/pkl-pantry/k8s.contrib.crd@1.0.0#/generate.pkl"

import "package://pkg.pkl-lang.org/pkl-k8s/k8s@1.0.1#/api/core/v1/ResourceRequirements.pkl"

source = "https://path/to/crd.yml"

converters {
  ["name.of.my.crd"] {
    [List("spec", "resources")] = ResourceRequirements
  }
}
```

Then, use this version via `pkl eval -m . generate.pkl`.

---

Old answer:

@jackkleeman is working on precisely this. This should be very similar to the [JSON Schema generator](https://github.com/apple/pkl-pantry/tree/main/packages/org.json_schema), except not quite, because of minor divergence from JSON Schema of OpenAPI

**Reply by Pythoner6** [reply]
Awesome! Yeah I figured it it'd be pretty similar to the existing code but saw that it was somewhat specific to the format of the k8s API json (e.g. dropping seven chars from names to get rid of an assumed `io.k8s.`).

Don't know if this is on the radar for the implementation but it'd be super neat if the generated pkl included CEL validations too.

**Reply by jackkleeman** [reply]
cel validations becoming pkl checks would be awesome for sure :D

**Comment by avarei** [comment]
Just saw that this is now available: https://github.com/apple/pkl-pantry/pull/17

**Reply by Pythoner6** [reply]
Thanks so much @jackkleeman for this, definitely excited to try this out! I did run into https://github.com/apple/pkl-pantry/issues/24 when trying it out on some real world crds I use, but I think that might be have a pretty simple fix https://github.com/apple/pkl-pantry/pull/25
